### PR TITLE
gh-111389: expose _PyHASH_INF/BITS/MODULUS/IMAG macros as public

### PR DESCRIPTION
**Pull request by skirpichev**
* Issue: gh-111389

### CODE REVIEW

**Comment by vstinner** [comment]
I would prefer to make them public if they are useful.

**Comment by skirpichev** [comment]
> I would prefer to make them public if they are useful.

@vstinner, are you about documenting these macroses somewhere (I propose https://docs.python.org/3/library/sys.html#sys.hash_info, where we could mention them in descriptions for different fields) AND about renaming them too (add PyUnstable_ aliases)?

File: Doc/library/sys.rst
Comment (vstinner) [inline Doc/library/sys.rst]: Usually we don't add references to the C API in the Python documentation. Moreover, usually we don't document the private API.

Comment (skirpichev) [reply Doc/library/sys.rst]: Redo in Doc/c-api/hashing.rst per [suggestion](https://github.com/python/cpython/issues/111389#issuecomment-1784940869) in the issue thread.

**Review by vstinner — CHANGES_REQUESTED** [review]
Please don't document private APIs.

**Comment by ngoldbaum** [comment]
It looks like numpy uses `_Py_HashDouble` to hash numpy scalars. If we're going to expose these symbols again, is there any appetite for exposing the functions used by them in the unstable API? If not it looks like it should be straightforward to vendor the code assuming there aren't any hidden issues for making the hash values of numpy scalars possibly differ from python scalars in the future.

**Comment by skirpichev** [comment]
> If you want to document these constants, make them public.

Ok, I've added PyUnstable prefixes.  If that's ok, I'll add a news entry.

The other option is to revert to 0ce377a.  I don't think we could just remove underscores, given the ``_PyHASH_NAN`` example.

I have made the requested changes; please review again

BTW, maybe we shouldn't export ``_PyHASH_BITS``.  This parameter is not available in ``sys.hash_info`` and examples for hashing in the docs show how to avoid using one.  I'll try to patch gmpy2 to remove it's usage (seems to be trivial for mpq, but may be less efficient for mpfr).  I think same could be done for the Sage.

> It looks like numpy uses _Py_HashDouble to hash numpy scalars. 

@ngoldbaum  FYI: https://github.com/python/cpython/issues/91744#issuecomment-1111990057

**Comment by skirpichev** [comment]
I have made the requested changes; please review again

**Comment by skirpichev** [comment]
I have made the requested changes; please review again

**Comment by skirpichev** [comment]
I didn't expect the Spanish Inquisition

**Comment by skirpichev** [comment]
@vstinner, not sure why bot isn't working, but [I think](https://github.com/python/cpython/pull/111418#issuecomment-1786556359) I did requested changes.

**Comment by vstinner** [comment]
Sorry, I was struggling with https://github.com/python/cpython/issues/111089 this week. I will try to go through all https://github.com/python/cpython/issues/111481 issues next week.

**Comment by skirpichev** [comment]
> BTW, maybe we shouldn't export _PyHASH_BITS. 

Hmm, no, it seems we really want this, e.g.: https://github.com/aleaxit/gmpy/pull/452 (see first removed commit).  Then maybe we should expose this in sys.hash_info struct too?

File: Include/pyhash.h
Comment (vstinner) [inline Include/pyhash.h]: Your change adds these API to the limited C API. I don't think that we want that.

I just created Include/cpython/pyhash.h to better distinguish the limited C API from the non-limited C API.

**Review by vstinner — COMMENTED** [review]
I'm working on PR https://github.com/python/cpython/pull/112095 which adds an "infrastructure" to test the PyHash API. I suggest waiting until this PR is merged before considering exposing these constants to the public C API.

File: Include/pyhash.h
Comment (skirpichev) [reply Include/pyhash.h]: > Your change adds these API to the limited C API. I don't think that we want that.

In past hashing of numeric types was changed in minor releases (3.x.0).  Are there reasons we don't want this freedom in future?

> I suggest waiting until this PR is merged before considering exposing these constants to the public C API.

Sound reasonable.

Comment (vstinner) [reply Include/pyhash.h]: > In past hashing of numeric types was changed in minor releases (3.x.0). Are there reasons we don't want this freedom in future?

It was proposed to create a C API Working Group. The Steering Council [has to take a decision](https://github.com/python/steering-council/issues/210). For now, no new API should be added to the limited C API.

See https://github.com/capi-workgroup/api-evolution/issues/41 and https://github.com/capi-workgroup/api-evolution/issues/42

Comment (skirpichev) [reply Include/pyhash.h]: Ok, added as public.  Docs moved to the hash.rst, news entry also added.  I don't think this could introduce merge conflicts with https://github.com/python/cpython/pull/112095

File: Misc/NEWS.d/next/C API/2023-11-15-09-24-51.gh-issue-111418.FYYetY.rst
Comment (vstinner) [inline Misc/NEWS.d/next/C API/2023-11-15-09-24-51.gh-issue-111418.FYYetY.rst]:
```suggestion
and :c:macro:`PyHASH_IMAG` C macros.  Patch by Sergey B Kirpichev.
```

File: Doc/c-api/hash.rst
Comment (vstinner) [inline Doc/c-api/hash.rst]: Please add `.. versionadded:: 3.13` to the added macros.

**Comment by skirpichev** [comment]
Perhaps, I should use ``Py_HASH_*`` naming convention? (Given #112449)

**Comment by skirpichev** [comment]
@vstinner, https://github.com/python/cpython/pull/112449 was closed.  On another hand, ``Py_HashPointer()`` function was added.  Does it make sense to keep this pr?

**Comment by skirpichev** [comment]
I have made the requested changes; please review again

File: Include/cpython/pyhash.h
Comment (vstinner) [inline Include/cpython/pyhash.h]: I would prefer to do the opposite: keep the old name as a deprecated alias, but define the public name to the value: 3 lines above. Same remark for other constants.

File: Doc/c-api/hash.rst
Comment (vstinner) [inline Doc/c-api/hash.rst]: I'm not sure that everybody knows about it. I suggest adding a link to Wikipedia:

```suggestion
   The `Mersenne prime <https://en.wikipedia.org/wiki/Mersenne_prime>`_ ``P = 2**n -1``, used for numeric hash scheme.
```

File: Doc/c-api/hash.rst
Comment (vstinner) [inline Doc/c-api/hash.rst]:
```suggestion
   The exponent ``n`` of ``P`` in :c:macro:`PyHASH_MODULUS`.
```

File: Include/cpython/pyhash.h
Comment (skirpichev) [reply Include/cpython/pyhash.h]: Ok, I did that.  Deprecated aliases added below in a separate block.

File: Include/cpython/pyhash.h
Comment (vstinner) [inline Include/cpython/pyhash.h:23]:
```suggestion
// Aliases kept for backward compatibility with Python 3.12
#define _PyHASH_BITS PyHASH_BITS
```

**Review by vstinner — APPROVED** [review]
LGTM, I just suggest to add a comment.

@mdickinson: Would you mind to double check this PR?

File: Include/cpython/pyhash.h
Comment (skirpichev) [reply Include/cpython/pyhash.h:23]: Done, but using ``/* style */`` for consistency with the rest of file.

**Review by mdickinson — APPROVED** [review]
Looks fine to me.

**Comment by vstinner** [comment]
Merged, thank you @skirpichev.

**Comment by skirpichev** [comment]
Thanks to all for review and patience.

**Pull request merged**
